src/java_bytecode/java_bytecode_convert_class.cpp

```diff
@@ -246,6 +246,11 @@ void java_bytecode_convert_classt::convert(
 /// java::array[X] where X is byte, float, int, char...
 void java_bytecode_convert_classt::add_array_types()
 {
+  // this method does not need to be called more than once
+  static bool already_executed=false;
+  if(already_executed)
+    return;
+
```

Comment:
I went back to a hybrid solution, as it indeed makes more sense if `add_array_types` is called from `operator()`, due to the fact that the code in `operator()` relies on this having been called before and that's the responsibility of `java_bytecode_convert_classt` rather than the user of `java_bytecode_convert_classt`.

Please have a quick look. I would need to merge this before 2pm.